resources/test262-agent-harness.js

```diff
@@ -0,0 +1,335 @@
+  iframe.src = URL.createObjectURL(blob);
+  document.body.appendChild(iframe);
+
+  let w = iframe.contentWindow;
+  // Finish test on completed event.
+  w.addEventListener('completed', t.step_func(function(e) {
+    t.done();
+  }));
+  // If test failed, rethrow error.
+  let FAILED = 'iframe-failed' + opts.strict ? 'strict' : '';
+  window.addEventListener(FAILED, t.step_func(function(e) {
+    t.set_status(t.FAIL);
+    throw new Error(e.detail);
+  }));
+  // In case of error send it to parent window.
+  w.addEventListener('error', function(e) { 
+    e.preventDefault();
+    // If the test failed due to a SyntaxError but phase was 'early', then the
+    // test should actually pass.
+    if (e.message.startsWith("SyntaxError")) {
+        if (attrs.type == 'SyntaxError' && attrs.phase == 'early') {
+            t.done();
+            return;
+        }
+    }
+    top.dispatchEvent(new CustomEvent(FAILED, {detail: e.message}));
+  });
+}
+
+let HEADER = `
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title></title>
+
+  <!-- Test262 harness -->
+  <script src="http://localhost:8000/resources/test262-harness.js"><\/script>
+
+  <!-- Test262 required libraries -->
+  <script src="http://localhost:8000/resources/test262/harness/assert.js"><\/script>
+  <script src="http://localhost:8000/resources/test262/harness/sta.js"><\/script>
+
+  ###INCLUDES###
+
+  <script type="text/javascript">
+    function __completed__() {
+      window.dispatchEvent(new CustomEvent('completed'));
+    }
+  <\/script>
+</head>
+<body>
+</body>
+<script type="text/javascript">
+`;
+let FOOTER = `
+  ;__completed__();
+<\/script>
+</html>
+`;
+
+function test262_as_html(test262, attrs, strict) {
+  output = [];
+  output.push(HEADER.replace('###INCLUDES###', addScripts(attrs.includes)));
+  output.push(prepareTest(test262, attrs, strict));
+  output.push(FOOTER);
+  return output.join("");
+}
+
+function addScripts(sources) {
+  sources = sources || [];
+  let ret = [];
+  let root = 'http://localhost:8000/resources/test262/harness/'
+  sources.forEach(function(src) {
+    ret.push("<script src='###SRC###'><\/script>".replace('###SRC###', root + src));
+  });
+  return ret.join("\n");
+}
+
+function prepareTest(test262, attrs, strict) {
+  function escapeEvalExpr(str) {
+      // Escape double quotes.
+      let ret = str.replace(/\"/g, '\\\"');
+      // Remove backreturn characters.
+      return ret.replace(/\n/g, '');
+  }
+  let output = [];
+  let test = test262();
+  if (strict) {
+    test = "'use strict';\n" + test;
+  }
+  if (attrs.type == 'SyntaxError' && attrs.phase == 'runtime') {
+      // If phase is 'runtime' the error will be caught here. If phase is 'early' the
+      // error will be handled at the window.onerror event.
+      output.push('try { eval("');
+      output.push(escapeEvalExpr(test));
+      output.push('"); }\ncatch (e) { assert.sameValue(e instanceof SyntaxError, true); }');
```

Comment:
The condition above will match tests containing the following:

```
negative:
  phase: runtime
  type: SyntaxError
```

The only tests that contain this particular set of attributes are those that test module code, direct eval and indirect eval: 

- [eval-code/direct/var-env-global-lex-non-strict.js](https://github.com/tc39/test262/tree/master/test/language/eval-code/direct/var-env-global-lex-non-strict.js)
- [eval-code/indirect/parse-failure-2.js](https://github.com/tc39/test262/tree/master/test/language/eval-code/indirect/parse-failure-2.js)
- [module-code/instn-iee-err-ambiguous-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-ambiguous-as.js)
- [module-code/instn-iee-err-ambiguous.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-ambiguous.js)
- [module-code/instn-iee-err-circular-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-circular-as.js)
- [module-code/instn-iee-err-circular.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-circular.js)
- [module-code/instn-iee-err-dflt-thru-star-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-dflt-thru-star-as.js)
- [module-code/instn-iee-err-dflt-thru-star.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-dflt-thru-star.js)
- [module-code/instn-iee-err-not-found-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-not-found-as.js)
- [module-code/instn-iee-err-not-found.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-iee-err-not-found.js)
- [module-code/instn-named-err-ambiguous-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-ambiguous-as.js)
- [module-code/instn-named-err-ambiguous.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-ambiguous.js)
- [module-code/instn-named-err-dflt-thru-star-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-dflt-thru-star-as.js)
- [module-code/instn-named-err-dflt-thru-star-dflt.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-dflt-thru-star-dflt.js)
- [module-code/instn-named-err-not-found-as.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-not-found-as.js)
- [module-code/instn-named-err-not-found-dflt.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-not-found-dflt.js)
- [module-code/instn-named-err-not-found.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-named-err-not-found.js)
- [module-code/instn-star-err-not-found.js](https://github.com/tc39/test262/tree/master/test/language/module-code/instn-star-err-not-found.js)

These can't be nested in a `try { eval(...)` }...`

Reply:
Here's a better illustration from firefox nightly:

This is code exactly as `prepareTest(...)` returns it: 

```js
try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
catch (e) { assert.sameValue(e instanceof SyntaxError, true); }
```

If I modify it to:

```js
try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
catch (e) { console.log(e); }
```

And put it in an html file, with the same `<script type="text/javascript">` (this is important) : 

```html
<!DOCTYPE html>
<script type="text/javascript">
  try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
  catch (e) { console.log(e); }
</script>
```

This will log: 

```
SyntaxError: import declarations may only appear at top level of a module
```


If I change that to (`<script type="module">` is the change): 

```js
<!DOCTYPE html>
<script type="module">
  try { eval("import x from './instn-named-err-not-found-empty_FIXTURE.js';"); }
  catch (e) { console.log(e); }
</script>
```

...The same error occurs. Also, in neither case does a network request get made for `instn-named-err-not-found-empty_FIXTURE.js`.


Finally, when I run this: 

```html
<!DOCTYPE html>
<script type="module">
import x from './instn-named-err-not-found-empty_FIXTURE.js';
</script>
```

The outcome is: 


```
SyntaxError: import not found: default
```

Which is the `SyntaxError` this test was written to verify. There is also a network request made for `instn-named-err-not-found-empty_FIXTURE.js`. (I did this whole exercise in Firefox Nightly with `dom.moduleScripts.enabled` set to `true` and Safari Technology Preview. 


------------------------------------------


Note: I had to manually make a file called `instn-named-err-not-found-empty_FIXTURE.js`, because the real `instn-named-err-not-found-empty_FIXTURE.js` has been turned into a file called `instn-named-err-not-found-empty_FIXTURE.html`, with all of the same treatments. Easily fixed by blacklisting `_FIXTURE.js` files.

Reply:
Nice catch! I will rework the code as suggested.